src/resolve/resolvectl.c

```diff
@@ -408,19 +408,49 @@ static int output_rr_packet(const void *d, size_t l, int ifindex) {
         return 0;
 }
 
+static int idna_candidate(const char *name, char **ret) {
+        _cleanup_free_ char *idnafied = NULL;
+        int r;
+
+        r = dns_name_apply_idna(name, &idnafied);
+        if (r < 0)
+                return log_error_errno(r, "Failed to apply IDNA to name '%s': %m", name);
+
+        if (r > 0 && !streq(name, idnafied)) {
+                *ret = TAKE_PTR(idnafied);
+                return true;
+        }
+
+        *ret = NULL;
+        return false;
+}
+
 static int resolve_record(sd_bus *bus, const char *name, uint16_t class, uint16_t type, bool warn_missing) {
         _cleanup_(sd_bus_message_unrefp) sd_bus_message *req = NULL, *reply = NULL;
         _cleanup_(sd_bus_error_free) sd_bus_error error = SD_BUS_ERROR_NULL;
+        _cleanup_free_ char *idnafied = NULL;
+        bool needs_authentication = false;
         unsigned n = 0;
         uint64_t flags;
-        int r;
         usec_t ts;
-        bool needs_authentication = false;
+        int r;
 
         assert(name);
 
         log_debug("Resolving %s %s %s (interface %s).", name, dns_class_to_string(class), dns_type_to_string(type), isempty(arg_ifname) ? "*" : arg_ifname);
 
+        if (dns_name_is_single_label(name))
+                log_notice("(Note that search domains are not appended when resolving raw record types. "
+                           "Please specify fully qualified domain names when resolving raw records, or remove --type= switch from invocation in order to request regular hostname resolution.)");
+
+        r = idna_candidate(name, &idnafied);
+        if (r < 0)
+                return r;
+        if (r > 0)
+                log_notice("(Note that IDNA translation is not applied when resolving raw record types. "
+                           "Please specify translated domain names — i.e. '%s' — when resolving raw records, or remove --type= switch from invocation in order to request regular hostname resolution.",
+                           idnafied);
```

Comment:
We don't know that. DNS actually doesn't actually mandate any specific encoding. The major TLDs/IANA have decided to adopt IDNA, but this is by no means mandatory, and people can actually put in their zones whatever they want, and bind and the other DNS servers will just serve it happily. Now, when resolving hostnames via classic DNS I think it's safe to just ignore all that, and as user-friendly operation just do the IDNA stuff and be done with it. But for raw RR record resolution I don't think we should do that, given that it's a lot more low-level: user supplies us with an exact query to make and we make it.

This is in particular relevant I think as neither mDNS nor LLMNR actually do the IDNA stuff. They just declared UTF-8 as the right encoding for hostnames, as-is. Handling the difference for high-level hostname lookups address is easy: we encode UTF-8 as IDNA when converting the suppied bus call paramaters when passing it to DNS but when passing it to LLMNR/mDNS we just pass it as is. But if we already get the properly prepared raw RR query info from the client, then it is weird to open up things again and remarshal it, I think.

This is even more complicated by the fact that IDNA exists in multiple versions. While the TLD zones insist on IDNA 2008 these days, zones further down can use older IDNA2003 (and they do, see #14483), and the rules are quite different for a number of characters. This means different encoding, and different allowed charcacters. There's a patch in this PR which enhances behaviour for this btw ("dns-domain: try IDN2003 rules if IDN2008 doesn't work"), i.e. where we try to be smart about it, and encode as IDN2003 if IDN2008 doesn't work, which is what web browsers do too. But it's nasty a bit, since it creates ambiguities: when doing a reverse lookup from address to such an IDNA domain, there are theoretically three different things we could return: the ASCII version of the domain name, the IDN2003 and the IDN2008 version... and they all look different in some cases. 

Hence, I think for low-level stuff (i.e. explicit, raw RR lookup, with `ResolveRecord()`) we should not try to be smart here, and put the burden on the user to do the translation. For the high-level stuff (i.e. domain→address and address→domain lookup with `ResolveHostname()` and `ResolveAddress()`) we should try to be smart and do all transformations as magically and sensibly as we possibly can, mimicing what web browsers do.

I hope that makes much sense?

Or to say this differently: I think it's a sensible thing that adding a `-t A` to a `resolvectl query` command moves you from "high-level", "user-friendly", "cooked" operation to "low-level", "raw" operation, where we do a lot less transformation in the background, giving the caller a lot more control about what is actually queried. People sholdn't use war RR lookups (`ResolveRecord()`…) if they just want to resolve an addres. They should use the proper APIs for that (`ResolveHostname()`…)

Reply:
I can see it both ways... But OK, let's do it like this, and if it turns out to be cumbersome in actual use, we can revisit.